File: cpp/0507.py
class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        i = 2
        sum = 1
        divisors = [1]

        if num <= 2:
            return False

        while i * i < num:
            if num % i == 0:
                if i not in divisors:
                    divisors.append(i)
                    sum += i
                if num / i not in divisors:
                    divisors.append(num / i)
                    sum += num / i

            i = i + 1
        
        return num == sum

File: cpp/0507.cpp
class Solution {
public:
    bool checkPerfectNumber(int num) {
        int sum = 1 + num;
        vector<int> divisors = { num, 1 };

        if (num == 1) {
            return false;
        }

        for (int i = 2; i < num; i++) {
            if (num % i == 0) {
                if ((num / i) == divisors.back()) {
                    break;
                }
                divisors.push_back(num / i);
                divisors.push_back(i);
                sum += (num / i);
                sum += i;
            }
        }
    
        return sum == num * 2;
    }
};
